File: sdk/csharp/src/AnyVali/Schemas/UnionSchema.cs
namespace AnyVali.Schemas;

public sealed class UnionSchema : Schema
{
    private readonly List<Schema> _variants;

    public UnionSchema(IEnumerable<Schema> variants)
    {
        _variants = variants.ToList();
    }

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        foreach (var variant in _variants)
        {
            var innerCtx = new ValidationContext();
            foreach (var p in ctx.Path)
                innerCtx.PushPath(p);

            var result = variant.RunPipeline(input, innerCtx);
            if (innerCtx.Issues.Count == 0)
                return result;
        }

        var variantKinds = _variants.Select(v => v.ToNode()["kind"]?.ToString() ?? "unknown");
        ctx.Issues.Add(new ValidationIssue
        {
            Code = IssueCodes.InvalidUnion,
            Message = "Input did not match any variant of the union",
            Path = ctx.ClonePath(),
            Expected = string.Join(" | ", variantKinds),
            Received = DescribeType(input),
        });
        return null;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?>
        {
            ["kind"] = "union",
            ["variants"] = _variants.Select(v => (object?)v.ToNode()).ToList(),
        };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new UnionSchema(_variants)
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}

File: sdk/ruby/lib/anyvali/schemas/union_schema.rb
# frozen_string_literal: true

module AnyVali
  class UnionSchema < Schema
    attr_reader :variant_schemas

    def initialize(variants:, **kwargs)
      @variant_schemas = variants.freeze
      super(kind: "union", **kwargs)
    end

    def to_node
      node = super
      node["variants"] = @variant_schemas.map(&:to_node)
      node
    end

    protected

    def validate(value, path, issues, context)
      @variant_schemas.each do |schema|
        result = schema.safe_parse(value, path: path, context: context)
        return if result.success?
      end

      # No variant matched
      expected = @variant_schemas.map(&:kind).join(" | ")
      issues << ValidationIssue.new(
        code: IssueCodes::INVALID_UNION,
        path: path,
        expected: expected,
        received: Schema.type_name(value)
      )
    end

    def dup_with(**overrides)
      variants = overrides.delete(:variants) || @variant_schemas
      attrs = {
        variants: variants,
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end
